Extra columns output for Kubectl get subcommand

mattmelgard opened the pull request:
/kind feature
/sig cli

**What this PR does / why we need it**:

This PR adds a new flag `--extra-columns` to `kubectl get` that will provide functionality similar to the currently implemented `--custom-columns` output flag but with the added feature that it prints the columns specified alongside the normal human-readable output of the command. 

Currently `--custom-columns` only supports printing those columns which are selected by the user via a comma-delimited list of `COLUMN_NAME` mapped to a jsonpath value (eg: `COLUMN:.metadata.name`) that results in something like this:

```
$ kubectl get pod my-pod --custom-columns=COLUMN_NAME=.metadata.name
COLUMN_NAME
my-pod
```

With this new flag, a user can expect to have the option to add these columns to the default set of columns to attain output like the following:

```
$ kubectl get pod my-pod --extra-columns=CUSTOM_COLUMN=.metadata.name
NAME     READY     STATUS    RESTARTS   AGE   CUSTOM_COLUMN
my-pod   1/1       Running   0          28s   my-pod
```

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubernetes/issues/71612

**Special notes for your reviewer**:
- Some minor refactoring was required to make `findJSONPathResults` callable from the `human-readable` printer, I moved this function into the `jsonpath` package to do this, but I'm open to other ways of implementing this if there are better places to put it.
- The same thing goes for the `Column` type that I found useful for this implementation. I moved this out of `customcolumn.go` and into `pkg/printers/interface.go`.
- I believe there are some scripts that need to be run in the hack directory to generate docs and a few other things, I still have not done that but I plan to add those changes soon
- I ran unit tests on the packages I changed, but I may still have broken some things, I will make sure to address any tests or packages that may have been affected by these changes
- Thanks for reviewing my first PR! I tried to follow the best practices and design patterns that were already present, but please do not hesitate to point out areas where I need to change things or where I might have missed some important parts of the contributor workflow. I'm still pretty new to both Golang and Kubernetes contributions.


```release-note
The kubectl get subcommand now supports appending additional columns to the normal get output through the --extra-columns flag
```

Commit:
Update tests for new extra-columns printing

mattmelgard commented:
/assign @soltysh

Commit:
Run make update

juanvallejo commented on pkg/kubectl/cmd/get/humanreadable_flags.go:
```diff
@@ -90,13 +94,31 @@ func (f *HumanPrintFlags) ToPrinter(outputFormat string) (printers.ResourcePrint
 		columnLabels = *f.ColumnLabels
 	}
 
+	spec := *f.ExtraColumns
```
nil check here please

juanvallejo commented on pkg/kubectl/cmd/get/humanreadable_flags.go:
```diff
@@ -90,13 +94,31 @@ func (f *HumanPrintFlags) ToPrinter(outputFormat string) (printers.ResourcePrint
+			if len(colSpec) != 2 {
+				return nil, fmt.Errorf("unexpected extra-columns spec: %s, expected <header>:<json-path-expr>", spec[ix])
+			}
+			spec, err := RelaxedJSONPathExpression(colSpec[1])
```
nit: since you're already using `spec` in the parent scope above, some potential confusion could be avoided here by choosing a slightly different name here

juanvallejo replied:
Also, consider putting this loop in its own function in this same file

juanvallejo commented on pkg/kubectl/cmd/get/humanreadable_flags_test.go:
```diff
@@ -44,6 +44,7 @@ func TestHumanReadablePrinterSupportsExpectedOptions(t *testing.T) {
 		// TODO(juanvallejo): test sorting once it's moved to the HumanReadablePrinter
 		sortBy       string
 		columnLabels []string
+		extraColumns []string
```
Since you're already in this file, consider adding a new test case here for extraColumns

juanvallejo replied:
Also, add a test case that tries adding an extraColumn with an "uncommon" value (intead of .metadata.name, for example, have the value be a .spec field)

mattmelgard replied:
Still working on adding some tests for these changes. Should have something in by tonight hopefully.

juanvallejo commented on staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go:
```diff
@@ -523,3 +525,11 @@ func (j *JSONPath) evalToText(v reflect.Value) ([]byte, error) {
 }
+
+// findJsonPath extracts a set of values from a runtime object using a parser
+func FindJSONPathResults(parser *JSONPath, from runtime.Object) ([][]reflect.Value, error) {
```
place `client-go` changes in their own commit please

juanvallejo commented on pkg/printers/humanreadable.go:
```diff
@@ -395,6 +408,86 @@ func PrintTable(table *metav1beta1.Table, output io.Writer, options PrintOptions
+// for wide columns and filtered rows and printing any extra columns using provided parsers. It filters out
+// rows that are Completed. You should call DecorateTable if you receive a table from a remote server before
+// calling PrintTableWithExtraCols.
+func PrintTableWithExtraCols(table *metav1beta1.Table, parsers []*jsonpath.JSONPath, output io.Writer, options PrintOptions) error {
```
We could try re-using most of the existing code in the `PrintTable` func by updating its signature to receive the new `parsers` parameter and handling that slice there, instead of creating this whole new function

juanvallejo commented:
/ok-to-test

mattmelgard commented:
@juanvallejo, thanks for the review, all good points which I'll be sure to address. As a side note, I didn't yet implement extra columns for the `-w` watch flag. Would I want to implement this feature for that as well?

Commit:
Extract findJSONPathResults into client-go jsonpath package

Commit:
Implement extra-columns in hr columns

Commit:
Add some changes suggested in review

- Break out json path expression parsing into its own function
- Add nil check before spec variable assignment
- Add spec parameter to PrintTable call signature and move
  functionality from PrintTableWithExtraCols function to that
  function instead

Commit:
Update references to PrintTable to use new call signature

mattmelgard commented:
Taking a look at these test failures now. I'm going to issue a retest in the meantime though since I don't think the e2e-gce test failures were as a result of my changes. /retest

mattmelgard commented:
/retest

fedebongio commented:
/assign @apelisse

Commit:
Add basic unit tests for extra columns

mattmelgard commented:
/retest

mattmelgard commented:
/test pull-kubernetes-kubemark-e2e-gce-big

soltysh commented on staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go:
```diff
@@ -523,3 +525,11 @@ func (j *JSONPath) evalToText(v reflect.Value) ([]byte, error) {
 	return buffer.Bytes(), nil
 }
+
+// findJsonPath extracts a set of values from a runtime object using a parser
```
I don't think this is desired here. I'm fine with duplicating that function in both places you need it. It's too small to bother making it a helper and this is definitely not a place for it.

soltysh commented on pkg/kubectl/cmd/get/customcolumn.go:
```diff
@@ -32,6 +32,7 @@ import (
 	"k8s.io/cli-runtime/pkg/printers"
 	"k8s.io/client-go/util/jsonpath"
 	utilprinters "k8s.io/kubernetes/pkg/kubectl/util/printers"
+	printerspkg "k8s.io/kubernetes/pkg/printers"
```
You can't introduce this dependency. We're in the process of splitting kubectl out of main repo and this is conflicting with that direction. I'd suggest leaving `Column` definition where it is.

soltysh commented on pkg/kubectl/cmd/get/get.go:
```diff
@@ -206,6 +207,13 @@ func (o *GetOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args []stri
 
 	sortBy, err := cmd.Flags().GetString("sort-by")
+
+	extraCols, err := cmd.Flags().GetStringSlice("extra-columns")
```
I think this should follow the current guideline where we specify `-o custom-columns=...` but I'll defer to @smarterclayton who was requesting that.

soltysh replied:
Alternatively we can support both.

soltysh commented on pkg/printers/humanreadable.go:
```diff
@@ -250,23 +250,20 @@ func (h *HumanReadablePrinter) PrintObj(obj runtime.Object, output io.Writer) er
-				if err := parsers[ix].Parse(columns[ix].FieldSpec); err != nil {
-					return err
-				}
+		columns := h.options.ExtraColumns
```
I'd expect this work similar to DecorateTable below by injecting the additional user-requested columns. It will make the code more readable and avoid printing logic duplication.

soltysh commented on pkg/printers/internalversion/printers_test.go:
```diff
@@ -1094,12 +1095,13 @@ func TestPrintHunmanReadableIngressWithColumnLabels(t *testing.T) {
 	}
 	verifyTable(t, table)
-	if err := printers.PrintTable(table, buff, printers.PrintOptions{NoHeaders: true}); err != nil {
+	if err := printers.PrintTable(table, parsers, buff, printers.PrintOptions{NoHeaders: true}); err != nil {
```
These will not be necessary when you apply the suggested change wrt decorating the table.

soltysh reviewed (changes requested):
Please address the review comments and once you do that re-arrange your commits since some of the will become obsolete. This will allow next review to be much cleaner.

mattmelgard commented:
/retest

heldic commented:
Is there any chance this gets picked up again? 
There seems to be a fair amount of interest on the [corresponding issue](https://github.com/kubernetes/kubernetes/issues/71612) (which has also been closed due to inactivity).
